test/plugin/test_metrics.rb

```diff
@@ -0,0 +1,294 @@
+
+      assert_false m.use_gauge_metric
+      assert_false m.has_methods_for_counter
+      assert_false m.has_methods_for_gauge
+    end
+
+    test 'all operations are not defined yet' do
+      assert_raise NotImplementedError do
+        @m.get('key')
+      end
+      assert_raise NotImplementedError do
+        @m.inc('key')
+      end
+      assert_raise NotImplementedError do
+        @m.dec('key')
+      end
+      assert_raise NotImplementedError do
+        @m.add('key', 10)
+      end
+      assert_raise NotImplementedError do
+        @m.sub('key', 11)
+      end
+      assert_raise NotImplementedError do
+        @m.set('key', 123)
+      end
+    end
+  end
+
+  sub_test_case 'BasicCounterMetric' do
+    setup do
+      @m = BasicCounterMetrics.new
+      @m.configure(config_element('metrics', '', {'@id' => '1'}))
+    end
+
+    test 'all operations work well' do
+      assert_true @m.has_methods_for_counter
+      assert_false @m.has_methods_for_gauge
+
+      assert_equal 0, @m.get('key')
+      assert_equal 1, @m.inc('key')
+
+      @m.add('key', 20)
+      assert_equal 21, @m.get('key')
+      assert_raise NotImplementedError do
+        @m.dec('key')
+      end
+
+      @m.set('key', 100)
+      assert_equal 100, @m.get('key')
+      assert_raise NotImplementedError do
+        @m.sub('key', 11)
+      end
+    end
+  end
+
+  sub_test_case 'AliasedCounterMetric' do
+    setup do
+      @m = AliasedCounterMetrics.new
+      @m.configure(config_element('metrics', '', {}))
+    end
+
+    test 'all operations work well' do
+      assert_true @m.has_methods_for_counter
+      assert_false @m.has_methods_for_gauge
+
+      assert_equal 0, @m.get('key')
+      assert_equal 1, @m.inc('key')
+
+      @m.add('key', 20)
+      assert_equal 21, @m.get('key')
+      assert_raise NotImplementedError do
+        @m.dec('key')
+      end
+
+      @m.set('key', 100)
+      assert_equal 100, @m.get('key')
+      assert_raise NotImplementedError do
+        @m.sub('key', 11)
+      end
+    end
+  end
+
+  sub_test_case 'BasicGaugeMetric' do
+    setup do
+      @m = BasicGaugeMetrics.new
+      @m.use_gauge_metric = true
+      @m.configure(config_element('metrics', '', {}))
+    end
+
+    test 'all operations work well' do
+      assert_false @m.has_methods_for_counter
+      assert_true @m.has_methods_for_gauge
+
+      assert_equal 0, @m.get('key')
+      assert_equal 1, @m.inc('key')
+
+      @m.add('key', 20)
+      assert_equal 21, @m.get('key')
+      @m.dec('key')
+      assert_equal 20, @m.get('key')
+
+      @m.set('key', 100)
+      assert_equal 100, @m.get('key')
+      @m.sub('key', 11)
+      assert_equal 89, @m.get('key')
+    end
+  end
+
+  sub_test_case 'AliasedGaugeMetric' do
+    setup do
+      @m = AliasedGaugeMetrics.new
+      @m.use_gauge_metric = true
+      @m.configure(config_element('metrics', '', {}))
+    end
+
+    test 'all operations work well' do
+      assert_false @m.has_methods_for_counter
```

Comment:
maybe better to mention dec set,sub

Reply:
#has_methods_for_gauge checks for set method definition and necessary methods.
Do you want to write down more descriptive comments or assertion, right?

Reply:
It is important whether the aliased method works as expected, so it may be better to focus on it.
another idea I mean:

```
test 'all common operations except aliased methods' do
...
end

test 'all aliased gauge operations' do
  assert_false @m.has_methods_for_counter
  ...
end
```

But it may be ok as is in the current way.